File: Backend/Backend/Models/Arqueo.cs
namespace Backend.Models;

public static class EstadoArqueo
{
    public const string Cuadrado = "CUADRADO";

    /// <summary>Se registró mal. No cuenta, pero se conserva con su rastro.</summary>
    public const string Anulado = "ANULADO";
}

/// <summary>
/// En qué se gastó el dinero de la ruta: pasaje, combustible, menú.
///
/// Es catálogo y no una lista fija de campos porque cada distribuidora gasta en
/// cosas distintas, y con campos fijos añadir "peaje" obligaría a migrar la
/// tabla y a tocar las tres capas.
/// </summary>
public class MotivoGasto
{
    public int Id { get; set; }
    public string Nombre { get; set; } = string.Empty;
    public string? Descripcion { get; set; }
    public bool Activo { get; set; } = true;
    public DateTime FechaCreacion { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// El cuadre de un día de reparto, de UNA persona.
///
/// Al volver de la ruta, quien cobró declara qué trae: el efectivo, en qué se
/// gastó parte por el camino, y los pagos digitales que le hicieron. El sistema
/// pone enfrente lo que dice que cobró esa persona ese día —las ventas al
/// contado que emitió y los abonos a deudas anteriores— y la diferencia salta
/// sola.
///
/// Es por persona y día, no uno global: en un cierre de toda la caja, un
/// faltante aparece igual pero sin saber de quién, que es justo lo que hay que
/// saber para resolverlo.
/// </summary>
public class ArqueoCaja
{
    public int Id { get; set; }

    /// <summary>El día que se cuadra (solo la fecha; la hora no importa).</summary>
    public DateTime Fecha { get; set; }

    /// <summary>De quién es el cuadre: quien salió a repartir y cobrar.</summary>
    public int UsuarioId { get; set; }
    public Usuario? Usuario { get; set; }

    // --- Lo que la persona declara ---

    public decimal Billetes { get; set; }
    public decimal Monedas { get; set; }

    // --- Lo que dicen los documentos, congelado al cerrar ---

    /// <summary>Lo que el sistema dice que cobró en efectivo ese día.</summary>
    public decimal EfectivoSistema { get; set; }

    /// <summary>Lo que el sistema dice que le entró por Yape, Plin o transferencia.</summary>
    public decimal BancosSistema { get; set; }

    public string? Observacion { get; set; }

    public string Estado { get; set; } = EstadoArqueo.Cuadrado;

    /// <summary>Quién lo registró, que puede no ser el mismo que cobró.</summary>
    public int? RegistradoPorId { get; set; }

    public DateTime FechaCreacion { get; set; } = DateTime.UtcNow;

    public List<ArqueoGasto> Gastos { get; set; } = [];
    public List<ArqueoPagoDigital> PagosDigitales { get; set; } = [];

    /// <summary>
    /// El efectivo que de verdad hubo en la mano: lo que entrega más lo que se
    /// gastó por el camino.
    ///
    /// Los gastos suman y no restan: ese dinero salió de lo cobrado, así que
    /// sin contarlo la ruta siempre parecería tener un faltante del tamaño de
    /// lo que se gastó en combustible.
    /// </summary>
    public decimal TotalEfectivoReal => Billetes + Monedas + Gastos.Sum(g => g.Monto);

    public decimal TotalDigitalReal => PagosDigitales.Sum(p => p.Monto);

    /// <summary>Negativa: falta dinero. Positiva: sobra.</summary>
    public decimal DiferenciaEfectivo => TotalEfectivoReal - EfectivoSistema;

    public decimal DiferenciaBancos => TotalDigitalReal - BancosSistema;

    /// <summary>
    /// Lo que la persona debe: el dinero que el sistema dice que cobró y no
    /// apareció. Es lo que se le descuenta.
    ///
    /// Se suman los faltantes de cada lado por separado, sin dejar que un
    /// sobrante compense: un Yape declarado que nunca llegó al banco es dinero
    /// perdido aunque ese día trajera de más en efectivo, y compensarlos
    /// escondería las dos cosas a la vez.
    /// </summary>
    public decimal Faltante =>
        (DiferenciaEfectivo < 0 ? -DiferenciaEfectivo : 0)
        + (DiferenciaBancos < 0 ? -DiferenciaBancos : 0);

    /// <summary>
    /// Lo que trajo de más. Solo se informa: no se le devuelve ni se le
    /// acumula a favor, porque un sobrante casi siempre es un cobro que se
    /// registró mal, no dinero suyo.
    /// </summary>
    public decimal Sobrante =>
        (DiferenciaEfectivo > 0 ? DiferenciaEfectivo : 0)
        + (DiferenciaBancos > 0 ? DiferenciaBancos : 0);

    /// <summary>
    /// El faltante ya se le descontó o lo repuso.
    ///
    /// Es una marca y no un borrado para que la deuda saldada siga siendo
    /// consultable: "este mes tuvo tres faltantes" es la información que
    /// importa cuando se repite.
    /// </summary>
    public bool FaltanteSaldado { get; set; }

    public DateTime? FechaSaldado { get; set; }
}

/// <summary>Un gasto de la ruta, descontado de lo que se trae.</summary>
public class ArqueoGasto
{
    public int Id { get; set; }

    public int ArqueoCajaId { get; set; }
    public ArqueoCaja? ArqueoCaja { get; set; }

    public int MotivoGastoId { get; set; }
    public MotivoGasto? MotivoGasto { get; set; }

    public decimal Monto { get; set; }

    /// <summary>Para el "otro" del catálogo, y para detallar cualquier gasto.</summary>
    public string? Descripcion { get; set; }
}

/// <summary>
/// Un cobro digital confirmado al cuadrar.
///
/// No se teclea: el sistema ya sabe qué cobró esa persona ese día, así que lo
/// que se hace es recorrer esos cobros y marcar cuáles llegaron de verdad.
/// Pedirle que los escriba de nuevo sería pedirle que los invente, y entonces
/// el cuadre no compara nada.
///
/// Lleva el número de operación porque es lo único que permite después
/// encontrarlo en el estado de cuenta del banco; sin él, un Yape que la
/// persona da por recibido y nunca llegó no se puede rastrear.
/// </summary>
public class ArqueoPagoDigital
{
    public int Id { get; set; }

    public int ArqueoCajaId { get; set; }
    public ArqueoCaja? ArqueoCaja { get; set; }

    /// <summary>
    /// El cobro del sistema que esta línea confirma.
    ///
    /// Nulo solo si ese cobro se borró después: la línea del cuadre se
    /// conserva igual, porque es el rastro de lo que se declaró aquel día.
    /// </summary>
    public int? PagoVentaId { get; set; }
    public PagoVenta? PagoVenta { get; set; }

    /// <summary>De quién vino. Nulo si no se supo identificar al cliente.</summary>
    public int? ClienteId { get; set; }
    public Cliente? Cliente { get; set; }

    public int MetodoPagoId { get; set; }
    public MetodoPago? MetodoPago { get; set; }

    public string? NumeroOperacion { get; set; }

    public decimal Monto { get; set; }
}

File: titanic/lib/features/finanzas/datos/arqueo.dart
// Modelos del cuadre de caja del reparto: calcan los DTOs del backend
// (ArqueoResponses.cs / ArqueoRequests.cs).

/// En que estado esta el cuadre de una persona en un dia.
class EstadoCuadre {
  const EstadoCuadre._();

  /// Cobro ese dia y todavia no ha cuadrado.
  static const pendiente = 'pendiente';

  /// Cuadro y no hubo diferencia.
  static const cuadrado = 'cuadrado';

  static const conDiferencia = 'conDiferencia';
  static const anulado = 'anulado';

  static String etiqueta(String estado) => switch (estado) {
    pendiente => 'Pendiente',
    cuadrado => 'Cuadrado',
    conDiferencia => 'Con diferencia',
    anulado => 'Anulado',
    _ => estado,
  };
}

double _num(dynamic v) => (v as num?)?.toDouble() ?? 0;

/// Motivo de gasto de la ruta: pasaje, combustible, menu.
class MotivoGasto {
  const MotivoGasto({
    required this.id,
    required this.nombre,
    required this.activo,
    required this.usos,
    this.descripcion,
  });

  final int id;
  final String nombre;
  final String? descripcion;
  final bool activo;
  final int usos;

  factory MotivoGasto.desdeJson(Map<String, dynamic> json) => MotivoGasto(
    id: json['id'] as int,
    nombre: json['nombre'] as String? ?? '',
    descripcion: json['descripcion'] as String?,
    activo: json['activo'] as bool? ?? true,
    usos: json['usos'] as int? ?? 0,
  );
}

/// Una fila de la lista: un dia y una persona.
///
/// Aparece aunque nadie haya cuadrado todavia, que es justo a quien hay que ir
/// a buscar.
class CuadrePendiente {
  const CuadrePendiente({
    required this.fecha,
    required this.usuarioId,
    required this.usuario,
    required this.efectivo,
    required this.bancos,
    required this.total,
    required this.estado,
    required this.faltante,
    required this.faltanteSaldado,
    this.diferenciaEfectivo,
    this.arqueoId,
  });

  final DateTime fecha;
  final int usuarioId;
  final String usuario;

  /// Lo que el sistema dice que cobro en efectivo ese dia.
  final double efectivo;

  /// Lo que le entro por Yape, Plin o transferencia.
  final double bancos;

  final double total;

  /// Solo si ya cuadro. Negativa: falta dinero.
  final double? diferenciaEfectivo;

  final String estado;

  /// El cuadre registrado, si lo hay: para corregirlo o anularlo.
  final int? arqueoId;

  final double faltante;
  final bool faltanteSaldado;

  bool get cuadrado => arqueoId != null && estado != EstadoCuadre.anulado;

  factory CuadrePendiente.desdeJson(Map<String, dynamic> json) =>
      CuadrePendiente(
        fecha: DateTime.parse(json['fecha'] as String),
        usuarioId: json['usuarioId'] as int,
        usuario: json['usuario'] as String? ?? '',
        efectivo: _num(json['efectivo']),
        bancos: _num(json['bancos']),
        total: _num(json['total']),
        diferenciaEfectivo: (json['diferenciaEfectivo'] as num?)?.toDouble(),
        estado: json['estado'] as String? ?? EstadoCuadre.pendiente,
        arqueoId: json['arqueoId'] as int?,
        faltante: _num(json['faltante']),
        faltanteSaldado: json['faltanteSaldado'] as bool? ?? false,
      );
}

/// Un cobro concreto: de quien vino y cuanto.
class CobroDelDia {
  const CobroDelDia({
    required this.pagoId,
    required this.fecha,
    required this.cliente,
    required this.documento,
    required this.metodoPagoId,
    required this.metodoPago,
    required this.tipoMetodo,
    required this.monto,
    required this.esDeudaAnterior,
  });

  final int pagoId;
  final DateTime fecha;
  final String cliente;

  /// El documento al que se aplico: NV-000012.
  final String documento;

  final int metodoPagoId;
  final String metodoPago;
  final String tipoMetodo;
  final double monto;

  /// El cobro salda una venta de otro dia: deuda vieja cobrada en la ruta.
  final bool esDeudaAnterior;

  factory CobroDelDia.desdeJson(Map<String, dynamic> json) => CobroDelDia(
    pagoId: json['pagoId'] as int? ?? 0,
    fecha: DateTime.parse(json['fecha'] as String),
    cliente: json['cliente'] as String? ?? '',
    documento: json['documento'] as String? ?? '',
    metodoPagoId: json['metodoPagoId'] as int? ?? 0,
    metodoPago: json['metodoPago'] as String? ?? '',
    tipoMetodo: json['tipoMetodo'] as String? ?? '',
    monto: _num(json['monto']),
    esDeudaAnterior: json['esDeudaAnterior'] as bool? ?? false,
  );
}

class ArqueoGasto {
  const ArqueoGasto({
    required this.id,
    required this.motivoGastoId,
    required this.motivoGasto,
    required this.monto,
    this.descripcion,
  });

  final int id;
  final int motivoGastoId;
  final String motivoGasto;
  final double monto;
  final String? descripcion;

  factory ArqueoGasto.desdeJson(Map<String, dynamic> json) => ArqueoGasto(
    id: json['id'] as int? ?? 0,
    motivoGastoId: json['motivoGastoId'] as int? ?? 0,
    motivoGasto: json['motivoGasto'] as String? ?? '',
    monto: _num(json['monto']),
    descripcion: json['descripcion'] as String?,
  );
}

class ArqueoPagoDigital {
  const ArqueoPagoDigital({
    required this.id,
    required this.metodoPagoId,
    required this.metodoPago,
    required this.monto,
    this.pagoVentaId,
    this.clienteId,
    this.cliente,
    this.numeroOperacion,
  });

  final int id;

  /// El cobro del sistema que esta linea confirma.
  final int? pagoVentaId;

  final int? clienteId;
  final String? cliente;
  final int metodoPagoId;
  final String metodoPago;
  final String? numeroOperacion;
  final double monto;

  factory ArqueoPagoDigital.desdeJson(Map<String, dynamic> json) =>
      ArqueoPagoDigital(
        id: json['id'] as int? ?? 0,
        pagoVentaId: json['pagoVentaId'] as int?,
        clienteId: json['clienteId'] as int?,
        cliente: json['cliente'] as String?,
        metodoPagoId: json['metodoPagoId'] as int? ?? 0,
        metodoPago: json['metodoPago'] as String? ?? '',
        numeroOperacion: json['numeroOperacion'] as String?,
        monto: _num(json['monto']),
      );
}

/// El cuadre ya declarado por una persona en un dia.
class ArqueoCaja {
  const ArqueoCaja({
    required this.id,
    required this.fecha,
    required this.usuarioId,
    required this.usuario,
    required this.billetes,
    required this.monedas,
    required this.efectivoSistema,
    required this.bancosSistema,
    required this.totalEfectivoReal,
    required this.totalDigitalReal,
    required this.diferenciaEfectivo,
    required this.diferenciaBancos,
    required this.faltante,
    required this.sobrante,
    required this.faltanteSaldado,
    required this.estado,
    required this.fechaCreacion,
    required this.gastos,
    required this.pagosDigitales,
    this.fechaSaldado,
    this.observacion,
    this.registradoPor,
  });

  final int id;
  final DateTime fecha;
  final int usuarioId;
  final String usuario;

  final double billetes;
  final double monedas;

  final double efectivoSistema;
  final double bancosSistema;

  final double totalEfectivoReal;
  final double totalDigitalReal;
  final double diferenciaEfectivo;
  final double diferenciaBancos;

  /// Lo que la persona debe reponer, si falto dinero.
  final double faltante;

  /// Lo que trajo de mas. Solo se informa.
  final double sobrante;

  final bool faltanteSaldado;
  final DateTime? fechaSaldado;

  final String? observacion;
  final String estado;
  final String? registradoPor;
  final DateTime fechaCreacion;

  final List<ArqueoGasto> gastos;
  final List<ArqueoPagoDigital> pagosDigitales;

  factory ArqueoCaja.desdeJson(Map<String, dynamic> json) => ArqueoCaja(
    id: json['id'] as int,
    fecha: DateTime.parse(json['fecha'] as String),
    usuarioId: json['usuarioId'] as int? ?? 0,
    usuario: json['usuario'] as String? ?? '',
    billetes: _num(json['billetes']),
    monedas: _num(json['monedas']),
    efectivoSistema: _num(json['efectivoSistema']),
    bancosSistema: _num(json['bancosSistema']),
    totalEfectivoReal: _num(json['totalEfectivoReal']),
    totalDigitalReal: _num(json['totalDigitalReal']),
    diferenciaEfectivo: _num(json['diferenciaEfectivo']),
    diferenciaBancos: _num(json['diferenciaBancos']),
    faltante: _num(json['faltante']),
    sobrante: _num(json['sobrante']),
    faltanteSaldado: json['faltanteSaldado'] as bool? ?? false,
    fechaSaldado: json['fechaSaldado'] == null
        ? null
        : DateTime.parse(json['fechaSaldado'] as String),
    observacion: json['observacion'] as String?,
    estado: json['estado'] as String? ?? '',
    registradoPor: json['registradoPor'] as String?,
    fechaCreacion: DateTime.parse(json['fechaCreacion'] as String),
    gastos: [
      for (final g in (json['gastos'] as List? ?? const []))
        ArqueoGasto.desdeJson(g as Map<String, dynamic>),
    ],
    pagosDigitales: [
      for (final p in (json['pagosDigitales'] as List? ?? const []))
        ArqueoPagoDigital.desdeJson(p as Map<String, dynamic>),
    ],
  );
}

/// Todo lo necesario para cuadrar a una persona en un dia.
class DetalleCuadre {
  const DetalleCuadre({
    required this.fecha,
    required this.usuarioId,
    required this.usuario,
    required this.efectivoSistema,
    required this.bancosSistema,
    required this.efectivo,
    required this.digital,
    this.arqueo,
  });

  final DateTime fecha;
  final int usuarioId;
  final String usuario;

  final double efectivoSistema;
  final double bancosSistema;

  /// Los cobros en efectivo, uno a uno: sin ver de quien se cobro no hay forma
  /// de encontrar el faltante.
  final List<CobroDelDia> efectivo;

  final List<CobroDelDia> digital;

  /// Lo declarado, si ya se cuadro.
  final ArqueoCaja? arqueo;

  factory DetalleCuadre.desdeJson(Map<String, dynamic> json) => DetalleCuadre(
    fecha: DateTime.parse(json['fecha'] as String),
    usuarioId: json['usuarioId'] as int? ?? 0,
    usuario: json['usuario'] as String? ?? '',
    efectivoSistema: _num(json['efectivoSistema']),
    bancosSistema: _num(json['bancosSistema']),
    efectivo: [
      for (final c in (json['efectivo'] as List? ?? const []))
        CobroDelDia.desdeJson(c as Map<String, dynamic>),
    ],
    digital: [
      for (final c in (json['digital'] as List? ?? const []))
        CobroDelDia.desdeJson(c as Map<String, dynamic>),
    ],
    arqueo: json['arqueo'] == null
        ? null
        : ArqueoCaja.desdeJson(json['arqueo'] as Map<String, dynamic>),
  );
}

/// Lo que una persona debe por faltantes, para descontarselo.
class DeudaUsuario {
  const DeudaUsuario({
    required this.usuarioId,
    required this.usuario,
    required this.pendiente,
    required this.dias,
    required this.saldado,
    required this.detalle,
  });

  final int usuarioId;
  final String usuario;

  /// Faltantes todavia sin saldar.
  final double pendiente;

  final int dias;

  /// Lo ya descontado o repuesto.
  final double saldado;

  final List<ArqueoCaja> detalle;

  factory DeudaUsuario.desdeJson(Map<String, dynamic> json) => DeudaUsuario(
    usuarioId: json['usuarioId'] as int? ?? 0,
    usuario: json['usuario'] as String? ?? '',
    pendiente: _num(json['pendiente']),
    dias: json['dias'] as int? ?? 0,
    saldado: _num(json['saldado']),
    detalle: [
      for (final a in (json['detalle'] as List? ?? const []))
        ArqueoCaja.desdeJson(a as Map<String, dynamic>),
    ],
  );
}
